### DISCUSSION TITLE
randombytes_buf  always generate 000000000000000000000000

### DISCUSSION THREAD

**Discussion by qijun4tian** [jedisct1/libsodium#1340] [Q&A]
hi 
i want encrypt in brower and write clang use crypto_aead_chacha20poly1305_ietf_encrypt , use randombytes_buf to generate npub  like below

```
 unsigned char n[crypto_aead_chacha20poly1305_IETF_NPUBBYTES];
    printf("\nnoce=");

    randombytes_buf(n, sizeof n);
    for (int i = 0; i < strlen(n); ++i){
        printf("%02x,", n[i]);
    }
    printf("\nout=");

    crypto_aead_chacha20poly1305_ietf_encrypt(out, &ciphertext_len,
                                              input, strlen(input),
                                              ADDITIONAL_DATA, ADDITIONAL_DATA_LEN,
                                              NULL, n, key);
```
and is work well on linux 
 so  i generate the wasm file like below`
emcc -O0  -s FORCE_FILESYSTEM=1  -s EXPORTED_FUNCTIONS='[_randombytes_buf,_main]' -s EXIT_RUNTIME=1  -I/root/libsodium-stable/src/libsodium/include /root/libsodium-stable/libsodium-js-sumo/lib/libsodium.a  main.c   -o output.html`
i run the output.html and randombytes_buf  always generate 000000000000000000000000
and also the brower console error 

`
Uncaught TypeError: Module.getRandomValue is not a function
`

**Comment by jedisct1** [comment]
The library must be initialized with "sodium_init()" before calling any other function. On all platforms.

**Reply by qijun4tian** [reply]
I call "sodium_init()" in main.c    and  crypto_aead_chacha20poly1305_ietf_encrypt work well

**Reply by qijun4tian** [reply]
```
  Module.onRuntimeInitialized = function() {
            Object.keys(_Module).forEach(function(k) {
              if (k !== 'getRandomValue') {
                delete _Module[k];
              }
            });
            Object.keys(Module).forEach(function(k) {
              _Module[k] = Module[k];
            });
            resolve();
          };
```

why delete getRandomValue  in emscripten.js

**Reply by jedisct1** [reply]
The `!==` operator means "different". So, that loops seems to delete all functions *except* `getRandomValue`.

**Comment by qijun4tian** [comment]
3ks. i recompile  it   and is ok now，  but  i do not know why is was not ok before.
